### PR TITLE
[Offload][OMPX] Add the runtime support for multi-dim grid and block

### PR DESCRIPTION
**Pull request by shiltian**

### CODE REVIEW

**Comment by shiltian** [comment]
* **#118042** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/118042?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/118042?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: offload/CMakeLists.txt [hunk 1]
```diff
@@ -218,6 +218,7 @@ if(LIBOMPTARGET_CMAKE_BUILD_TYPE MATCHES debug)
 else()
   option(LIBOMPTARGET_ENABLE_DEBUG "Allow debug output with the environment variable LIBOMPTARGET_DEBUG=1" OFF)
 endif()
+option(LIBOMPTARGET_ENABLE_DEBUG "Allow debug output with the environment variable LIBOMPTARGET_DEBUG=1" ON)
```
Comment (shiltian) [inline offload/CMakeLists.txt:221]: Hmm, this is for debug only...

**Comment by jplehr** [comment]
Out of curiosity: Have you looked at how this applies downstream?

**Comment by shiltian** [comment]
> Out of curiosity: Have you looked at how this applies downstream?

No, I didn't.

File: offload/plugins-nextgen/common/src/PluginInterface.cpp [hunk 2]
```diff
@@ -616,38 +622,36 @@ KernelLaunchParamsTy GenericKernelTy::prepareArgs(
...
-uint32_t GenericKernelTy::getNumThreads(GenericDeviceTy &GenericDevice,
-                                        uint32_t ThreadLimitClause[3]) const {
-  assert(ThreadLimitClause[1] == 0 && ThreadLimitClause[2] == 0 &&
-         "Multi dimensional launch not supported yet.");
-
-  if (IsBareKernel && ThreadLimitClause[0] > 0)
-    return ThreadLimitClause[0];
+void GenericKernelTy::getNumThreads(GenericDeviceTy &GenericDevice,
+                                    uint32_t ThreadLimitClause[3]) const {
+  if (IsBareKernel)
```
[the remark this answers is not in the document]
Comment (shiltian) [reply offload/plugins-nextgen/common/src/PluginInterface.cpp:627]: No, the logic is different because we directly use whatever users provide for us in bare mode.

File: offload/plugins-nextgen/common/src/PluginInterface.cpp [hunk 3]
```diff
@@ -566,10 +567,26 @@ Error GenericKernelTy::launch(GenericDeviceTy &GenericDevice, void **ArgPtrs,
...
-      getNumBlocks(GenericDevice, KernelArgs.NumTeams, KernelArgs.Tripcount,
-                   NumThreads, KernelArgs.ThreadLimit[0] > 0);
+  uint32_t NumThreads[3] = {KernelArgs.ThreadLimit[0],
+                            KernelArgs.ThreadLimit[1],
+                            KernelArgs.ThreadLimit[2]};
+  uint32_t NumBlocks[3] = {KernelArgs.NumTeams[0], KernelArgs.NumTeams[1],
+                           KernelArgs.NumTeams[2]};
+  // FIXME: This is a WA to "calibrate" the bad work done in the front end.
+  // Delete this ugly code after the front end emits proper values.
+  auto CorrectMultiDim = [](uint32_t (&Val)[3]) {
```
Comment (arsenm) [inline offload/plugins-nextgen/common/src/PluginInterface.cpp:577]: Just directly fix this? Where is this coming from?

**Comment by jplehr** [comment]
I ran this through our buildbot config and the new tests do not seem to pass. The 0-th dimension appears to be the culprit. Can you double check that it works?
If you need any help / info, let me know.

```
# .---command stderr------------
# | llvm-project/offload/test/api/omp_env_vars.c:8:12: error: CHECK: expected string not found in input                                                                                                       
# |  // CHECK: Launching kernel [[KERNEL:.+_main_.+]] with [1,1,1] blocks and [1,1,1] threads
# |            ^
# | <stdin>:1:1: note: scanning from here
# | "PluginInterface" device 0 info: Launching kernel __omp_offloading_fd00_61c4b8b_main_l9 with [0,1,1] blocks and [0,1,1] threads in Generic mode                                                                                     
# | ^
# | <stdin>:1:55: note: possible intended match here
# | "PluginInterface" device 0 info: Launching kernel __omp_offloading_fd00_61c4b8b_main_l9 with [0,1,1] blocks and [0,1,1] threads in Generic mode                                                                                     
# |                                                       ^
# |
# | Input file: <stdin>
# | Check file: llvm-project/offload/test/api/omp_env_vars.c
# |
# | -dump-input=help explains the following input dump.
# |
# | Input was:
# | <<<<<<
# |            1: "PluginInterface" device 0 info: Launching kernel __omp_offloading_fd00_61c4b8b_main_l9 with [0,1,1] blocks and [0,1,1] threads in Generic mode                                                                       
# | check:8'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found                                                
# | check:8'1                                                           ?                                                                                          possible intended match                                              
# | >>>>>>
# `-----------------------------
# error: command failed with exit status: 1
```

**Comment by shiltian** [comment]
@jplehr I tested it locally, but couldn't reproduce the failure.

```
➜ LIBOMPTARGET_INFO=63 ./main
omptarget device 0 info: Entering OpenMP kernel at unknown:0:0 with 9 arguments:
omptarget device 0 info: from(unknown)[49152]
omptarget device 0 info: to(unknown)[49152]
omptarget device 0 info: to(unknown)[49152]
omptarget device 0 info: firstprivate(unknown)[4] (implicit)
omptarget device 0 info: firstprivate(unknown)[4] (implicit)
omptarget device 0 info: firstprivate(unknown)[4] (implicit)
omptarget device 0 info: firstprivate(unknown)[4] (implicit)
omptarget device 0 info: firstprivate(unknown)[4] (implicit)
omptarget device 0 info: firstprivate(unknown)[4] (implicit)
omptarget device 0 info: Creating new map entry with HstPtrBase=0x000063ac60a736b0, HstPtrBegin=0x000063ac60a736b0, TgtAllocBegin=0x000079d64fa00000, TgtPtrBegin=0x000079d64fa00000, Size=49152, DynRefCount=1, HoldRefCount=0, Name=unknown
omptarget device 0 info: Creating new map entry with HstPtrBase=0x000063ac60a32500, HstPtrBegin=0x000063ac60a32500, TgtAllocBegin=0x000079d64fa0c000, TgtPtrBegin=0x000079d64fa0c000, Size=49152, DynRefCount=1, HoldRefCount=0, Name=unknown
omptarget device 0 info: Copying data from host to device, HstPtr=0x000063ac60a32500, TgtPtr=0x000079d64fa0c000, Size=49152, Name=unknown
omptarget device 0 info: Creating new map entry with HstPtrBase=0x000063ac60a676a0, HstPtrBegin=0x000063ac60a676a0, TgtAllocBegin=0x000079d64fa18000, TgtPtrBegin=0x000079d64fa18000, Size=49152, DynRefCount=1, HoldRefCount=0, Name=unknown
omptarget device 0 info: Copying data from host to device, HstPtr=0x000063ac60a676a0, TgtPtr=0x000079d64fa18000, Size=49152, Name=unknown
omptarget device 0 info: Mapping exists with HstPtrBegin=0x000063ac60a736b0, TgtPtrBegin=0x000079d64fa00000, Size=49152, DynRefCount=1 (update suppressed), HoldRefCount=0
omptarget device 0 info: Mapping exists with HstPtrBegin=0x000063ac60a32500, TgtPtrBegin=0x000079d64fa0c000, Size=49152, DynRefCount=1 (update suppressed), HoldRefCount=0
omptarget device 0 info: Mapping exists with HstPtrBegin=0x000063ac60a676a0, TgtPtrBegin=0x000079d64fa18000, Size=49152, DynRefCount=1 (update suppressed), HoldRefCount=0
"PluginInterface" device 0 info: Launching kernel __omp_offloading_10302_8ee2def_main_l28 with [2,4,6] blocks and [32,4,2] threads in SPMD mode
AMDGPU device 0 info: #Args: 10 Teams x Thrds:   48x 256 (MaxFlatWorkGroupSize: 1024) LDS Usage: 0B #SGPRs/VGPRs: 16/6 #SGPR/VGPR Spills: 0/0 Tripcount: 0
omptarget device 0 info: Mapping exists with HstPtrBegin=0x000063ac60a676a0, TgtPtrBegin=0x000079d64fa18000, Size=49152, DynRefCount=0 (decremented, delayed deletion), HoldRefCount=0
omptarget device 0 info: Mapping exists with HstPtrBegin=0x000063ac60a32500, TgtPtrBegin=0x000079d64fa0c000, Size=49152, DynRefCount=0 (decremented, delayed deletion), HoldRefCount=0
omptarget device 0 info: Mapping exists with HstPtrBegin=0x000063ac60a736b0, TgtPtrBegin=0x000079d64fa00000, Size=49152, DynRefCount=0 (decremented, delayed deletion), HoldRefCount=0
omptarget device 0 info: Copying data from device to host, TgtPtr=0x000079d64fa00000, HstPtr=0x000063ac60a736b0, Size=49152, Name=unknown
omptarget device 0 info: Removing map entry with HstPtrBegin=0x000063ac60a676a0, TgtPtrBegin=0x000079d64fa18000, Size=49152, Name=unknown
omptarget device 0 info: Removing map entry with HstPtrBegin=0x000063ac60a32500, TgtPtrBegin=0x000079d64fa0c000, Size=49152, Name=unknown
omptarget device 0 info: Removing map entry with HstPtrBegin=0x000063ac60a736b0, TgtPtrBegin=0x000079d64fa00000, Size=49152, Name=unknown
omptarget device 0 info: OpenMP Host-Device pointer mappings table empty
omptarget device 0 info: OpenMP Host-Device pointer mappings table empty
```

**Comment by shiltian** [comment]
@jplehr I tested it locally but couldn't reproduce the failure locally.

```
➜  clang -fopenmp --offload-arch=gfx1030 llvm-project/offload/test/api/omp_env_vars.c -o main
➜  env OMP_NUM_TEAMS=1 OMP_TEAMS_THREAD_LIMIT=1 LIBOMPTARGET_INFO=16 ./main
"PluginInterface" device 0 info: Launching kernel __omp_offloading_10302_8ee073e_main_l9 with [1,1,1] blocks and [1,1,1] threads in SPMD mode
AMDGPU device 0 info: #Args: 1 Teams x Thrds:    1x   1 (MaxFlatWorkGroupSize: 256) LDS Usage: 56B #SGPRs/VGPRs: 39/32 #SGPR/VGPR Spills: 0/0 Tripcount: 0
```

File: offload/test/offloading/ompx_bare_multi_dim.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
+// RUN: env LIBOMPTARGET_INFO=63 %libomptarget-compilexx-run-and-check-generic
```
Comment (jplehr) [inline offload/test/offloading/ompx_bare_multi_dim.cpp:1]: Are you sure that this env propagates to the `run` stage of the test too?

I think this is why I see an error when running the botconfig. I do not see an error when I run the binary manually. Other tests appear to split this command into two. Example from `test/offloading/force-usm.cpp`

```
// clang-format off
// RUN: %libomptarget-compilexx-generic
// RUN: env LIBOMPTARGET_INFO=32 %libomptarget-run-generic 2>&1 | %fcheck-generic --check-prefix=NO-USM
```

Comment (shiltian) [reply offload/test/offloading/ompx_bare_multi_dim.cpp:1]: That's a really weird limitation...

**Review by jdoerfert — APPROVED** [review]
LG, fix the test.
Consider not printing dim 2 and 3 if we run in OpenMP mode, but OK for now.

**Comment by shiltian** [comment]
@jplehr Is that the only failure you observed? Could you help check the latest change?

**Comment by jplehr** [comment]
> @jplehr Is that the only failure you observed? Could you help check the latest change?

Let me give this another shot. I'll let you know!

**Comment by jplehr** [comment]
I see these two failures on a local test run
``` 
  libomptarget :: x86_64-unknown-linux-gnu :: api/omp_env_vars.c
  libomptarget :: x86_64-unknown-linux-gnu-LTO :: api/omp_env_vars.c
```

The error message is
```
# .---command stderr------------
# | /work/janplehr/git/llvm-project/offload/test/api/omp_env_vars.c:8:12: error: CHECK: expected string not found in input
# |  // CHECK: Launching kernel [[KERNEL:.+_main_.+]] with [1,1,1] blocks and [1,1,1] threads
# |            ^
...
# | Input was:
# | <<<<<<
# |            1: "PluginInterface" device 0 info: Launching kernel __omp_offloading_fd00_61c4b8b_main_l9 with [0,1,1] blocks and [0,1,1] threads in Generic mode 
# | check:8'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
# | check:8'1                                                           ?                                                                                          possible intended match
# | >>>>>>
```

**Comment by shiltian** [comment]
I ended up with forcing `api/omp_env_vars.c` to use GPU.

**Review by jplehr — APPROVED** [review]
Ran this through local testing and saw the issues resolved.

File: offload/plugins-nextgen/common/src/PluginInterface.cpp (hunk 3 above)
Comment (shiltian) [reply offload/plugins-nextgen/common/src/PluginInterface.cpp]: This is from the front end. In regular OpenMP, 0 represents users don't specify any value such that the runtime needs to choose one. The multi-dim kernel launch is purely an extension and it has not been standardized yet. I'll do a follow up patch that emits `[0,1,1]` for regular OpenMP if nothing is specified and `[x,y,z]` for OMPX.

**Comment by ronlieb** [comment]
> > Out of curiosity: Have you looked at how this applies downstream?
> 
> No, I didn't.

in hindsight, i wish you had

**Comment by jhuber6** [comment]
> in hindsight, i wish you had

This pain is very much self-inflicted I'm afraid

File: offload/plugins-nextgen/common/src/PluginInterface.cpp [hunk 5]
```diff
@@ -566,10 +567,16 @@ Error GenericKernelTy::launch(GenericDeviceTy &GenericDevice, void **ArgPtrs,
...
-  uint32_t NumThreads = getNumThreads(GenericDevice, KernelArgs.ThreadLimit);
-  uint64_t NumBlocks =
-      getNumBlocks(GenericDevice, KernelArgs.NumTeams, KernelArgs.Tripcount,
-                   NumThreads, KernelArgs.ThreadLimit[0] > 0);
+  uint32_t NumThreads[3] = {KernelArgs.ThreadLimit[0],
+                            KernelArgs.ThreadLimit[1],
+                            KernelArgs.ThreadLimit[2]};
+  uint32_t NumBlocks[3] = {KernelArgs.NumTeams[0], KernelArgs.NumTeams[1],
+                           KernelArgs.NumTeams[2]};
+  if (!IsBareKernel) {
```
Comment (dhruvachak) [inline offload/plugins-nextgen/common/src/PluginInterface.cpp:575]: Before this patch, IsBareKernel was not checked. Is this change required for the multi-dim support? Or is it unrelated change? Same question for other new uses of IsBareKernel.

Comment (shiltian) [reply offload/plugins-nextgen/common/src/PluginInterface.cpp:575]: Before this patch the check is in the two get functions.
